F-FrontEnd/src/run_tests.py

```diff
@@ -0,0 +1,782 @@
+from enum import Enum, IntEnum
+from typing import Tuple, NamedTuple, List, Optional, Dict, Union
+from textwrap import wrap
+from contextlib import contextmanager
+import datetime as dt
+import test_report as tr
+from lxml import etree
+
+THIS_DIR_PATH = os.path.dirname(os.path.realpath(__file__))
+TEST_DATA_DEFAULT_RELATIVE_PATH = '../../F-FrontEnd/test/testdata'
+DEFAULT_ERROR_LOG_FILENAME = 'errors.log'
+TMPFS_DIR = '/dev/shm'
+TEST_TIMEOUT = 10  # seconds
+TEST_REPORT_SCHEMA_FILE = join_path(THIS_DIR_PATH, 'test-report-schema.xsd')
+DEFAULT_VIEW_XSLT = 'test-report-view.xslt'
+
+
+def file_exists(path: str):
+    return os.path.exists(path) and os.path.isfile(path)
+
+
+def dir_exists(path: str):
+    return os.path.exists(path) and os.path.isdir(path)
+
+
+def get_file_as_str(path: str):
+    with open(path, 'r') as f:
+        s = f.read().replace('\n', ' ').strip()
+        return s
+
+
+class NativeCompiler(Enum):
+    GFORTRAN = 1
+    PGFORTRAN = 2
+    UNKNOWN = 3
+
+
+class TesterArgs(NamedTuple):
+    num_parallel_tests: int
+    frontend: str
+    backend: str
+    xmodules_dir: str
+    native_compiler_type: NativeCompiler
+    native_compiler: str
+    errors_log: str
+    test_data_dir: str
+    test_case: Optional[str]
+    working_dir: Optional[str]
+    verbose_output: bool
+    obj_sym_reader: str
+    jobs_num: int
+    break_on_failed: bool
+    report_file: str
+
+
+class TestCaseStageID(Enum):
+    DEPENDENCIES_PREP = 0
+    FRONTEND = 1
+    BACKEND = 2
+    NATIVE = 3
+    LINK = 4
+    SYMBOLS_READ = 5
+    EXECUTION = 6
+    EXPECTED_OUTPUT = 7
+    REFERENCE_OUTPUT = 8
+
+
+class DT(IntEnum):
+    US = 1
+    S = 1000000 * US
+    D = 24 * 60 * 60 * S
+
+    @staticmethod
+    def us_duration(ts1: dt.datetime, ts2: dt.datetime):
+        d = ts2 - ts1
+        duration = d.microseconds * DT.US + d.seconds * DT.S + d.days * DT.D
+        return duration
+
+
+class TestCaseStageResult(NamedTuple):
+    type: TestCaseStageID
+    result: bool
+    args: Optional[str]
+    error_log: Optional[str]
+    start_timestamp: dt.datetime
+    end_timestamp: dt.datetime
+
+    def to_xml_data(self):
+        duration = DT.us_duration(self.start_timestamp, self.end_timestamp)
+        data = tr.TestCaseStage(start_timestamp=self.start_timestamp, end_timestamp=self.end_timestamp,
+                                duration=duration, result=self.result, name=self.type.name)
+        if self.args is not None:
+            data.set_arguments(self.args)
+        if self.error_log is not None:
+            data.set_error_log(self.error_log)
+        return data
+
+
+class TerminalColor:
+    BLACK = '\033[30m'
+    RED = '\033[31m'
+    GREEN = '\033[32m'
+
+
+def to_paragraph(p: Union[str, bytes], prefix: str = ''):
+    if isinstance(p, bytes):
+        p = p.decode('utf-8')
+    return '\n'.join([prefix + l for l in wrap(p, width=100)])
+
+
+def error_text(s: str):
+    return TerminalColor.RED + s + TerminalColor.BLACK
```

Comment:
Not sure if it will do the right thing, but `'\033[m'` should _reset_ the colors.

>  `ESC[m` acts like a `0` reset code

Source: https://en.wikipedia.org/wiki/ANSI_escape_code#CSI_sequences

(`\033` = `ESC` in octal format)

Reply:
Correct, added the switch to default color.